pkg/agent/controller/networkpolicy/packetin.go

```diff
@@ -278,17 +271,62 @@ func (c *Controller) rejectRequest(pktIn *ofctrl.PacketIn) error {
 	binary.BigEndian.PutUint32(icmpData[:ICMPUnusedHdrLen], 0)
 	copy(icmpData[ICMPUnusedHdrLen:], ipHdr[:ipHdrLen+8])
 	return c.ofClient.SendICMPPacketOut(
-		srcMAC.String(),
-		dstMAC.String(),
+		srcMAC,
+		dstMAC,
 		srcIP,
 		dstIP,
-		inPort,
-		-1,
+		uint32(inPort),
+		outPort,
 		isIPv6,
 		icmpType,
 		icmpCode,
 		icmpData,
-		true)
+		packetOutType)
+}
+
+// isServiceTraffic uses ServiceEPStateField to decided if this is a Service
+// traffic, when AntreaProxy is enabled. And when AntreaProxy is disabled, if the
+// IP of reject response packet is on local Node, but the MAC is MAC of antrea-gw,
+// then this is a rejection of Service traffic. Because this response is heading to
+// kube-proxy.
+func (c *Controller) isServiceTraffic(pktIn *ofctrl.PacketIn, antreaProxyEnabled bool, dstMAC string, isDstLocal bool) bool {
+	if antreaProxyEnabled {
+		matches := pktIn.GetMatches()
+		match := getMatchRegField(matches, openflow.ServiceEPStateField)
+		if match != nil {
+			svcEpstate, err := getInfoInReg(match, openflow.ServiceEPStateField.GetRange().ToNXRange())
+			return err == nil && svcEpstate&openflow.EpSelectedRegMark.GetValue() == openflow.EpSelectedRegMark.GetValue()
+		}
+	}
+	gwIfaces := c.ifaceStore.GetInterfacesByType(interfacestore.GatewayInterface)
+	return isDstLocal && dstMAC == gwIfaces[0].MAC.String()
+}
+
+// getRejectOFPort sets the inPort and outPort of a packetOut based on the packetOutType.
+func getRejectOFPort(packetOutType binding.PacketOutType, sIface, dIface *interfacestore.InterfaceConfig) (int32, int32) {
+	inPort := int32(config.HostGatewayOFPort)
+	outPort := int32(-1)
```

Comment:
I would prefer to use uint32 for the `outPort` field in a PacketOut message, otherwise it should be difficult to be compatible with some special port numbers like "Normal"/"Local" action in future. And you can use "0" to ignore the output port number.

Reply:
Changed.